cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py

```diff
@@ -0,0 +1,227 @@
+    DeveloperErrorViewMixin,
+    view_auth_classes,
+)
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+
+# TODO: Potential future view.
+# @view_auth_classes(is_authenticated=True)
+# class DownstreamListView(DeveloperErrorViewMixin, APIView):
+#     """
+#     List all blocks which are linked to upstream content, with optional filtering.
+#     """
+#     def get(self, request: Request) -> Response:
+#         """
+#         Handle the request.
+#         """
+#         course_key_string = request.GET['course_id']
+#         syncable = request.GET['ready_to_sync']
+#         ...
+
+
+@view_auth_classes(is_authenticated=True)
+class DownstreamView(DeveloperErrorViewMixin, APIView):
+    """
+    Inspect or manage an XBlock's link to upstream content.
+    """
+    def get(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Inspect an XBlock's link to upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=False)
+        _ensure_upstream_ref(downstream)
+        if link := UpstreamLink.try_get_for_block(downstream):
+            return Response(link.to_json())
+        raise ValidationError(detail=f"Block '{usage_key_string}' is not linked to an upstream")
+
+    def put(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Edit an XBlock's link to upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+        new_upstream_ref = request.data.get("upstream_ref")
+        if not isinstance(new_upstream_ref, str):
+            raise ValidationError({"upstream_ref": "value missing"})
+
+        # Set `downstream.upstream` so that we can try to sync and/or fetch.
+        # Note that, if this fails and we raise a 4XX, then we will not call modulstore().update_item,
+        # thus preserving the former value of `downstream.upstream`.
+        downstream.upstream = new_upstream_ref
+        sync_param = request.data.get("sync", "false").lower()
+        if sync_param not in ["true", "false"]:
+            raise ValidationError({"sync": "must be 'true' or 'false'"})
+        try:
+            if sync_param == "true":
+                sync_from_upstream(downstream=downstream, user=request.user)
+            else:
+                fetch_customizable_fields(downstream=downstream, user=request.user)
+        except BadDownstream as exc:
+            raise ValidationError(str(exc)) from exc
+        except BadUpstream as exc:
+            raise ValidationError({"upstream_ref": str(exc)}) from exc
+        modulestore().update_item(downstream, request.user.id)
+        link = UpstreamLink.get_for_block(downstream)
+        assert link
```

Comment:
This was also to make mypy happy-- it wanted us to check for None since `get_for_block` will return None iff `downstream.upstream is None`, but we've ascertained above that `downstream.upstream` is non-None, so it's a spurious type warning.

I changed it in my latest commit to just use a `# type: ignore[assignment]` rather than an assert.

Reply:
Based on [these changes](https://github.com/openedx/edx-platform/pull/34925#discussion_r1801586472), I was able to make `get_for_block` return `UpstreamLink` (non-nullable), which made this type-ignore stuff unneccessary.